src/main/scala/com/wavesplatform/network/HistoryReplier.scala

```diff
@@ -6,70 +6,69 @@ import com.wavesplatform.settings.SynchronizationSettings
 
-  // both cache loaders will throw a NoSuchElementException when (micro)block is absent, but it's fine, because:
-  // 1. cache reads are wrapped in a Future
-  // 2. neither MicroBlockRequest nor GetBlock require a response when the requested (micro)block can not be found
+  private implicit val scheduler: SchedulerService = monix.execution.Scheduler.fixedPool(
+    name = "history-replier",
+    reporter = com.wavesplatform.utils.UncaughtExceptionsToLogReporter,
+    poolSize = 2
+  )
 
   private val knownMicroBlocks = CacheBuilder.newBuilder()
     .maximumSize(historyReplierSettings.maxMicroBlockCacheSize)
     .build(new CacheLoader[MicroBlockSignature, Array[Byte]] {
       override def load(key: MicroBlockSignature) =
-        blocking(history.microBlock(key))
+        history.microBlock(key)
           .map(m => MicroBlockResponseMessageSpec.serializeData(MicroBlockResponse(m))).get
     })
 
   private val knownBlocks = CacheBuilder.newBuilder()
     .maximumSize(historyReplierSettings.maxBlockCacheSize)
     .build(new CacheLoader[ByteStr, Array[Byte]] {
-      override def load(key: ByteStr) = blocking(history.heightOf(key).flatMap(history.blockBytes)).get
+      override def load(key: ByteStr) = history.heightOf(key).flatMap(history.blockBytes).get
     })
 
   override def channelRead(ctx: ChannelHandlerContext, msg: AnyRef): Unit = msg match {
-    case GetSignatures(otherSigs) => Future {
+    case GetSignatures(otherSigs) => Task {
       otherSigs.view
-        .map(parent => parent -> blocking(history.blockIdsAfter(parent, settings.maxChainLength)))
+        .map(parent => parent -> history.blockIdsAfter(parent, settings.maxChainLength))
         .find(_._2.nonEmpty) match {
         case Some((parent, extension)) =>
           log.debug(s"${id(ctx)} Got GetSignatures with ${otherSigs.length}, found common parent $parent and sending ${extension.length} more signatures")
           ctx.writeAndFlush(Signatures(parent +: extension))
-        case None if otherSigs.length == 1 && otherSigs.head == blocking(history.lastBlock).get.uniqueId =>
+        case None if otherSigs.length == 1 && otherSigs.head == history.lastBlock.get.uniqueId =>
           // this is the special case when both nodes only have genesis block
           log.debug(s"${id(ctx)} Both local and remote nodes only have genesis block")
           ctx.writeAndFlush(Signatures(otherSigs))
         case _ =>
           log.debug(s"${id(ctx)} Got GetSignatures with ${otherSigs.length} signatures, but could not find an extension")
       }
-    }
+    }.runAsync
 
-    case GetBlock(sig) =>
-      Future(knownBlocks.get(sig)).foreach { bytes =>
-        ctx.writeAndFlush(RawBytes(BlockMessageSpec.messageCode, bytes))
-      }
+    case GetBlock(sig) => Task {
+      Option(knownBlocks.get(sig)).foreach(bytes =>
+        ctx.writeAndFlush(RawBytes(BlockMessageSpec.messageCode, bytes)))
+    }.runAsync
```

Comment:
`knownBlocks` is a concurrent data structure. Also we can't send a signature of block we didn't processed. So, the remote node should not ask it.